src/Microsoft.NetCore.Analyzers/Core/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -0,0 +1,330 @@
+                    !wellKnownTypeProvider.TryGetTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreMvcFiltersIFilterMetadata, out var iFilterMetadataTypeSymbol) ||
+                    !wellKnownTypeProvider.TryGetTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreAntiforgeryIAntiforgery, out var iAntiforgeryTypeSymbol) ||
+                    !wellKnownTypeProvider.TryGetTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreMvcFiltersIAsyncAuthorizationFilter, out var iAsyncAuthorizationFilterTypeSymbol) ||
+                    !wellKnownTypeProvider.TryGetTypeByMetadataName(WellKnownTypeNames.SystemThreadingTasksTask, out var taskTypeSymbol) ||
+                    !wellKnownTypeProvider.TryGetTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreMvcFiltersAuthorizationFilterContext, out var authorizationFilterContextTypeSymbol))
+                {
+                    return;
+                }
+
+                // A dictionary from method symbol to set of methods calling it directly.
+                // The bool value in the sub ConcurrentDictionary is not used, use ConcurrentDictionary rather than HashSet just for the concurrency security.
+                var inverseGraph = new ConcurrentDictionary<ISymbol, ConcurrentDictionary<ISymbol, bool>>();
+
+                // Ignore cases where a global anti forgery filter is in use.
+                var hasGlobalAntiForgeryFilter = false;
+
+                // Verify that validate anti forgery token attributes are used somewhere within this project,
+                // to avoid reporting false positives on projects that use an alternative approach to mitigate CSRF issues.
+                var usingValidateAntiForgeryAttribute = false;
+                var onAuthorizationAsyncMethodSymbols = new HashSet<IMethodSymbol>();
+                var actionMethodSymbols = new HashSet<(IMethodSymbol, string)>();
+
+                // Constructing inverse callGraph.
+                // When it comes to delegate function assignment Del handler = DelegateMethod;, inverse call Graph will add:
+                // (1) key: method gets called in DelegateMethod, value: handler.
+                // When it comes to calling delegate function handler(), inverse callGraph will add:
+                // (1) key: delegate function handler, value: callerMethod.
+                // (2) key: Invoke(), value: callerMethod.
+                compilationStartAnalysisContext.RegisterOperationBlockStartAction(
+                    (OperationBlockStartAnalysisContext operationBlockStartAnalysisContext) =>
+                    {
+                        if (hasGlobalAntiForgeryFilter)
+                        {
+                            return;
+                        }
+
+                        var owningSymbol = operationBlockStartAnalysisContext.OwningSymbol;
+                        inverseGraph.TryAdd(owningSymbol, new ConcurrentDictionary<ISymbol, bool>());
```

Comment:
>inverseGraph.TryAdd(owningSymbol, new ConcurrentDictionary<ISymbol, bool>()); [](start = 24, length = 77)

Avoid instantiating a new ConcurrentDictionary in every call.

`inverseGraph.GetOrAdd(owningSymbol, (_) => new ConcurrentDictionary<ISymbol, bool>());`